py/dml/crep.py

```diff
@@ -19,8 +19,52 @@
     'ctype',
     'conf_obj',
     'cloggroup',
+    'dev',
+    'require_dev',
+    'maybe_dev',
+    'maybe_dev_arg',
+    'StaticContext',
+    'TypedParamContext',
     )
 
+
+class StaticContext:
+    active = False
```

Comment:
I would find it more natural to invert the default, i.e. we evaluate in static context by default, and enter "instance context" when evaluating a method body. I.e., if we enter a context where an instance is known, then that makes `dev()` return a reference to that instance. Perhaps even softcode the name, `with InstanceContext('_dev')`, to eliminate the need for `maybe_dev_arg`.

Reply:
That'll be a pain. I'll look into it, but I'd like to keep the static context approach if it proves too difficult.

Reply:
Partly done. I've done a straight-up inversion (`StaticContext` to `DeviceInstanceContext´), but nothing else, yet -- I'm confused about the rest of your comment, and want to talk about it. In the meantime, the conversion is messy and likely fragile, so it warrants an additional test against VP.